utils/src/server/rpc_module.rs

```diff
@@ -31,22 +29,18 @@ pub type MethodSink = mpsc::UnboundedSender<String>;
 type Subscribers = Arc<Mutex<FxHashMap<(ConnectionId, SubscriptionId), (MethodSink, oneshot::Receiver<()>)>>>;
 
 /// Sets of JSON-RPC methods can be organized into a "module"s that are in turn registered on the server or,
-/// alternatively, merged with other modules to construct a cohesive API.
-#[derive(Default)]
-pub struct RpcModule {
+/// alternatively, merged with other modules to construct a cohesive API. [`RpcModule`] wraps an additional context
+/// argument that can be used to access data during call execution.
+pub struct RpcModule<Context> {
+	ctx: Arc<Context>,
 	methods: Methods,
 	subscribers: Subscribers,
 }
 
-impl RpcModule {
-	/// Instantiate a new `RpcModule`.
-	pub fn new() -> Self {
-		RpcModule { methods: Methods::default(), subscribers: Subscribers::default() }
-	}
-
-	/// Add context for this module, turning it into an `RpcContextModule`.
-	pub fn with_context<Context>(self, ctx: Context) -> RpcContextModule<Context> {
-		RpcContextModule { ctx: Arc::new(ctx), module: self, subscribers: Subscribers::default() }
+impl<Context> RpcModule<Context> {
+	/// Create a new module with a given shared `Context`.
+	pub fn new(ctx: Context) -> Self {
```

Comment:
@niklasad1 Having to add a context everywhere is what got me thinking about why the *module* needs a context when in reality it's actually an attribute of a *call*.

Reply:
yes, that's a good observation I think it's by-design to avoid having to clone the `Context` on each call.

Otherwise, you either have to borrow (sounds tricky) or move (take by value) context in the each register_method.

For register_subscription it's assumed that the `context` will be moved into a thread or task and need to be Arc:ed or impl Clone.  However, we could let the user decide that an just require `Context` there but might be hard understand the compiler errors...

Reply:
I reckon being able to put some state into your callbacks is a pretty common use case. It's not strictly necessary, but without it you'd have to provide your own state (likely by doing what we are doing internally with `Arc<Context>`), which could be very boiler plate'y.

That said, if we can swipe all that boiler plate under the rug with proc macro instead, we could potentially remove the context altogether. I do like the `RpcModule` not having generics for ease of use too.